apollo-federation/src/sources/connect/fetch_dependency_graph/mod.rs

```diff
@@ -93,10 +96,231 @@ impl FetchDependencyGraphApi for FetchDependencyGraph {
+        // Note: Comparison of two numbers is fast
+        if source_path.source_entering_edge != source_data.source_entering_edge {
+            return Err(FederationError::internal(
+                "expected to have matching entering edges",
+            ));
+        }
+
+        // If we don't have any field selections in the source_path, then there is nothing to do.
+        let Some(source_path_field) = source_path.field else {
+            return Ok(());
+        };
+
+        // Enforce that the field contains properties shared with the source data
+        if source_path_field.response_name != source_data.field_response_name {
+            return Err(FederationError::internal(
+                "expected path and source data to have the same field name",
+            ));
+        }
+        if source_path_field.arguments != source_data.field_arguments {
+            return Err(FederationError::internal(
+                "expected path and source data to have the same field arguments",
+            ));
+        }
+
+        // Ensure that we have a selection, inserting an initial value if not.
+        let selection =
+            source_data
+                .selection
+                .get_or_insert_with(|| match &source_path_field.selections {
+                    // Construct a new selection from the supplied path properties
+                    PathSelections::Selections {
+                        head_property_path,
+                        tail_selection,
+                        ..
+                    } => {
+                        if head_property_path.is_empty() {
+                            JSONSelection::Named(SubSelection::default())
+                        } else if let Some((_name, _tail)) = tail_selection {
+                            JSONSelection::Path(PathSelection::from_slice(
+                                head_property_path,
+                                Some(SubSelection::default()),
+                            ))
+                        } else {
+                            JSONSelection::Path(PathSelection::from_slice(head_property_path, None))
+                        }
+                    }
+
+                    // Pass through the supplied selection
+                    PathSelections::CustomScalarRoot { selection } => selection.clone(),
+                });
+
+        // TODO: Matching twice seems sad, but how can we separate the selection logic from the traversal?
+        // If figured out, remove clone above
+        if let PathSelections::Selections {
+            named_selections,
+            tail_selection,
+            ..
+        } = source_path_field.selections
+        {
+            // We can short out if there's nothing to select
+            let Some((tail_name, tail_subselection)) = tail_selection else {
+                return Ok(());
+            };
+
+            // If we are adding a path and have a selection, then the selection _must_ have a subselection.
```

Comment:
i'd love to have examples in the comments like you have in tests, is this `"foo" + "{ bar }" = "foo { bar }"`?

Reply:
Somewhat done!